BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -103,16 +103,13 @@ void BSP_CAN_Init(void) {
 /**
  * @brief   Transmits the data onto the CAN bus with the specified id
  * @param   id : Message of ID. Also indicates the priority of message. The lower the value, the higher the priority.
- * @param   data : data to be transmitted. The max is 8 bytes.
- * @param   length : num of bytes of data to be transmitted. This must be <= 8 bytes or else the rest of the message is dropped.
+ * @param   Voltage : the voltage value at that moment in a float
  * @return  0 if module was unable to transmit the data onto the CAN bus. Any other value indicates data was transmitted.
  */
-uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
+uint8_t BSP_CAN_Write(uint32_t id, float Voltage) {
     TxMessage.StdId = id;
-    TxMessage.DLC = length;
-	for(int i = 0; i < length; i++){
-        TxMessage.Data[i] = data[i];
-	}
+    TxMessage.StdId = Voltage;
```

Comment:
You are just overwriting the CAN ID here. Look at the definition of the `CanTxMsg` struct in `stm32f4xx_can.h` to see the different fields in this struct

Reply:
You are setting the `StdId` field of the struct `TxMessage` struct twice. See the code below (copy pasted from `BSP/STM32F413/STM32F4xx_StdPeriph_Driver/Inc/stm32f4xx_can.h` for information on what the different fields in the struct are.

```C
/** 
  * @brief  CAN Tx message structure definition  
  */
typedef struct
{
  uint32_t StdId;  /*!< Specifies the standard identifier.
                        This parameter can be a value between 0 to 0x7FF. */

  uint32_t ExtId;  /*!< Specifies the extended identifier.
                        This parameter can be a value between 0 to 0x1FFFFFFF. */

  uint8_t IDE;     /*!< Specifies the type of identifier for the message that 
                        will be transmitted. This parameter can be a value 
                        of @ref CAN_identifier_type */

  uint8_t RTR;     /*!< Specifies the type of frame for the message that will 
                        be transmitted. This parameter can be a value of 
                        @ref CAN_remote_transmission_request */

  uint8_t DLC;     /*!< Specifies the length of the frame that will be 
                        transmitted. This parameter can be a value between 
                        0 to 8 */

  uint8_t Data[8]; /*!< Contains the data to be transmitted. It ranges from 0 
                        to 0xFF. */
} CanTxMsg;
```